aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -324,201 +309,120 @@ impl Aggregator {
-                            }
-                            err
-                        })?;
-                        DeltaValue::Positive(
-                            addition(current_delta, input, self.max_value)
-                                .expect("Addition is expected to succeed as the sum <= max_value"),
-                        )
-                    },
-                    DeltaValue::Negative(current_delta) => {
-                        // If speculative_start_value - current_delta + input is greater than max_value or less than -max_value, it's an overflow.
-                        // Otherwise, we update the delta with delta + input.
-                        if speculative_start_value.get_value() + input < current_delta {
-                            // speculative_start_value - current_delta + input < 0
-                            DeltaValue::Negative(subtraction(current_delta, input).expect(
-                                "Subtraction of smaller value from larger value must succeed",
-                            ))
-                        } else {
-                            // speculative_start_value - current_delta + input > 0
-                            if speculative_start_value.get_value() + input - current_delta
-                                > self.max_value
-                            {
-                                self.get_mut_history()
-                                    .unwrap()
-                                    .record_overflow(input - current_delta);
-                                return Err(abort_error(
-                                    format!("overflow occurred when adding {} to speculative_start_value: {:?}, current_delta: -{}", input, speculative_start_value, current_delta),
-                                    EADD_OVERFLOW,
-                                ));
-                            }
-                            if input > current_delta {
-                                DeltaValue::Positive(subtraction(input, current_delta).expect(
-                                    "Subtraction of smaller value from larger value must succeed",
-                                ))
-                            } else {
-                                DeltaValue::Negative(subtraction(current_delta, input).expect(
-                                    "Subtraction of smaller value from larger value must succeed",
-                                ))
-                            }
-                        }
-                    },
-                };
-                self.state = AggregatorState::Delta {
-                    speculative_start_value,
-                    delta: new_delta,
-                    history,
-                };
-                self.get_mut_history().unwrap().record_success(new_delta);
+                let cur_value = expect_ok(addition_deltavalue(speculative_start_value.get_value(), delta, self.max_value))?;
```

Comment:
we should be able to simplify this by computing current_value first.

also I think it would be useful to distinguish things that should never happen (which expect_ok here does) and we are not fataling purely because we are in move, from the errors that mean speculative execution is incorrect and needs to be re-executed.

i.e. here, unless we have a bug in the code, we should always be able to add delta to speculative start value